lib/puppet/type/ios_interface.rb

```diff
@@ -0,0 +1,46 @@
+require 'puppet/resource_api'
+
+Puppet::ResourceApi.register_type(
+  name: 'ios_interface',
+  docs: 'Manage layer 3 configuration on a per Instance basis.',
+  features: ['remote_resource'],
+  attributes: {
+    name: {
+      type:      'String',
+      desc:      'The switch interface name, e.g. "Ethernet1"',
+      behaviour: :namevar,
+    },
+    mac_notification_added: {
+      type:    'Optional[Boolean]',
+      desc:    'Whether to enable Mac Address added notification for this port.',
+    },
+    mac_notification_removed: {
+      type:    'Optional[Boolean]',
+      desc:    'Whether to enable Mac Address removed notification for this port.',
+    },
+    link_status: {
+      type:    'Optional[Boolean]',
+      desc:    'Whether to permit duplicate SNMP LINKUP and LINKDOWN traps.',
+    },
+    logging_event: {
+      type:   'Optional[Array[Enum["bundle-status","nfas-status","spanning-tree","status","subif-link-status","trunk-status"]]]',
+      desc:   'Whether or not to log certain event messages. Any event log not specifically indicated will be disabled.',
```

Comment:
Is there a reason to disable if `logging_event` is not included? As if something is not in the manifest then it should be unmanaged, i.e. if they do not supply a `logging_event` then I don't think the current logging_events should be removed unless there is a particular reason with some docs we can refer them too?

Reply:
If the code was set up that way there'd be no way to set all logging_event's too false.

Reply:
I know it isn't pretty, but the alternatives aren't any better.

Reply:
You could use the `unset` word which is used in others, so on the `get` if no `logging_events` exist then `unset` is returned, and then to `unset` the user supplies `unset`. The current implementation goes against how Puppet is meant to work and unmanaged (i.e. user hasn't supplied anything) should not be touched in any form

Reply:
That could work, if they enter unset then compare the list of current against the the list of should and unset any differences.
Will implement now.

Reply:
Have implemented as I said above, unfortunately it causes idempotency issues that can't be solved by having an unset value. 
i.e. logging_event => ['trunk-status','unset']
Not sure how to get around them.

Reply:
Well this is a combination of things that you seem to be trying to achieve. 

i.e. a manifest:

```
ios_interface { 'ethernet1/1':
logging_event => ['trunk-status'],
}
```

This would mean this is the only thing I wish to be set and remove existing values. The issue with the original implementation was that if the user didn't supply anything relating to `logging_event` then any logging event on the device would have been removed which wouldn't be ideal and goes against Puppet not meant to manage things that you don't specify - what I meant by `unset` was to just have an enum of `unset` supported so that if someone supplied:

```
ios_interface { 'ethernet1/1':
logging_event => 'unset',
}
```

Then this would remove the existing `logging_events`. so instead of having it like 

```
      type:   'Optional[Array[Enum["unset","bundle-status","nfas-status","spanning-tree","status","subif-link-status","trunk-status"]]]',
```

It would be 

```
      type:   'Optional[Variant[Enum["unset"], Array[Enum["bundle-status","nfas-status","spanning-tree","status","subif-link-status","trunk-status"]]]]',
```

So they can either supply `unset` or the previous array of `logging_events`

Reply:
Didn't think of that. Will implement it now.

Reply:
Implemented :)